arendst/Tasmota
# Allow non-admin user handling on web interface

## Description:

**Related issue (if applicable):** fixes #5719, option iii)

## Checklist:
  - [x] The pull request is done against the latest dev branch
  - [x] Only relevant files were touched (Also remember to update _changelog.ino_ file)
  - [x] Only one feature/fix was added per PR.
  - [x] The code change is tested and works.
  - [x] The code change pass travis tests. **Your PR cannot be merged unless tests pass**
  - [x] I accept the [CLA](https://github.com/arendst/Sonoff-Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

## Impact:

* This change adds 208 bytes to "sonoff.bin".
* Added a new variable to settings, hence the minor version change from ...9 to ...10.

## How to test:
(note on "restart browser" below: that is needed to flush the user/password. ALL browser windows need to be closed. Basic Authentication is primitive.. )

* Set an admin web password (/configuration/other configuration/web admin password), but leave user password empty
* Restart browser
* Go to device main page
* TEST: the main page should be limited in functionality now, close to how "Webserver 1" behaves ( although now one can get back to full functionality via the web alone.)
* Click on the "configuration" button, or request any other valid page (/co, /cm,...)
* TEST: it should ask for a login now
* Enter "admin" and the password set for "admin"
* TEST: access should be given to full functionality now.
* Set the user password
* Restart browser
* Go to device main page
* TEST: it should ask for a password now
* Enter "user" and the password set for "user"
* TEST: the main page should be limited in functionality now, like above.
* Click on the "configuration" button, or request any other valid page (/co, /cm,...)
* TEST: it should ask for a login now. If you do that too slow, the browser may say "401", as that is the result of the home page refresh functionality intervening. Just Reload then.
* Enter "admin" and the password set for "admin"
* TEST: access should be given to full functionality now.
* Execute "webserver 1" command (MQTT or web command) and let the device reset
* Restart browser
* Go to device main page
* Enter "user" and the password set for "user"
* TEST: Only limited functionality should be available now. 
* Restart browser
* Go to device main page
* Enter "admin" and the password set for "admin"
* TEST: Only limited functionality should be available now. 
* Optional: if you want to go back to "webserver 2" mode, do so via MQTT or a full reflash.
* Do a fresh install of the device, with WiFi setup and all.
* TEST: that should be unchanged and work without problems

## Things that might need improvement:

* Translations for  D_WEB_USER_PASSWORD (especially German, not happy with that translation)
* When upgrading, the "user" password is empty. One might not want that. Should I set it the same as the admin password?
* Explanations in the wiki about how this works (I can do that, once the PR has been approved), especially 
  * the user names "admin" and "user", that are hard coded, and are mostly not hinted at in the various translations.
  * "webserver 1": that mode cannot be reversed via the web interface (nothing changed with that, but it is not noted in the Wiki, and might be confusing). 

## Other remarks:

* for this PR, I did a large cleanup of all places where authentication was done, as there were code and functionality duplications.
* There is no "UI 'Lock down' button", as was mentioned in the ticket. Not needed. Just the additional "user" profile that was mentioned.

## New functionality:

* Two users can be used and be assigned their own password:
  * admin: access to all functions.
  * user: access only to the home page, and interaction with devices. No configuration, no firmware updates.
* When upgrading from older firmware, the user password is set as empty, meaning: no authentication is needed for user (non-admin) level access. NOTE: This may not be wanted. 
* Modified home page and config screen, see below, and see explanations of the screens.
* The "UserPassword" command has been added, acting on the user password, behaving like the "WebPassword" command that acts on the admin password.
* "webserver 1"/"HTTP_USER" mode is accessible to both admin and user. 

## Screen captures:

**Home page, "admin" logged in** (as before):
<img width="383" alt="Screenshot 2019-05-05 at 17 48 34" src="https://user-images.githubusercontent.com/12572422/57196550-39de7500-6f5e-11e9-8282-56897466c9d0.png">

**Home page, "user" logged in**:
<img width="383" alt="Screenshot 2019-05-05 at 17 49 08" src="https://user-images.githubusercontent.com/12572422/57196560-5f6b7e80-6f5e-11e9-8d5a-fb457eff2395.png">
The bottom button will ask for the admin password, as the "user" has no access to configuration. That button is there so that admins have an easy way to log in as "admin". Please note that the authentication method used by Tasmota (Basic Authentication) does not allow web server control over logging out; it is al controlled by the browser.

**Other Configuration screen**:
<img width="383" alt="Screenshot 2019-05-05 at 19 27 41" src="https://user-images.githubusercontent.com/12572422/57197809-030f5b80-6f6c-11e9-9495-41f1391e8bbb.png">
Note the added "User password" field.
